File: plugins/geo_fencing/android/src/main/kotlin/com/example/geo_fencing/GeoFencingPlugin.kt
package com.example.geo_fencing

import android.content.Context
import com.example.geo_fencing.actions.result.FailureResult
import com.example.geo_fencing.actions.result.SuccessResult
import com.example.geo_fencing.factory.PluginMethodFactory
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result

/** GeoFencingPlugin */
class GeoFencingPlugin : FlutterPlugin, MethodCallHandler {

    companion object {
        const val INIT_METHOD_NAME = "init_geo_fencing"
        const val INSERT_METHOD_NAME ="insert_location"
    }

    private lateinit var pluginMethodFactory: PluginMethodFactory
    private lateinit var channel: MethodChannel
    private var context: Context? = null
    override fun onAttachedToEngine(flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
        channel = MethodChannel(flutterPluginBinding.binaryMessenger, "ir.vhdrjb/geo_fencing")
        channel.setMethodCallHandler(this)
        pluginMethodFactory = PluginMethodFactory()
        context = flutterPluginBinding.applicationContext
    }

    override fun onMethodCall(call: MethodCall, result: Result) {
        when (val methodResult = pluginMethodFactory.createMethodAction(call, context!!)) {
            is FailureResult -> {
                result.error(methodResult.error, null, null)
            }

            is SuccessResult -> {
                result.success(methodResult.result)
            }

            else -> {
                result.notImplemented()
            }
        }
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel.setMethodCallHandler(null)
        context = null
    }
}

File: plugins/geo_fencing/ios/Classes/GeoFencingPlugin.swift
import Flutter
import UIKit

public class GeoFencingPlugin: NSObject, FlutterPlugin {
    
  
    static let INIT_METHOD_NAME = "init_geo_fencing"
    static let INSERT_METHOD_NAME = "insert_location"
    
    public static func register(with registrar: FlutterPluginRegistrar) {
        let channel = FlutterMethodChannel(name: "ir.vhdrjb/geo_fencing", binaryMessenger: registrar.messenger())
        let instance = GeoFencingPlugin()
        registrar.addMethodCallDelegate(instance, channel: channel)
    }
    
    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let methodAction = PluginMethodFactory().createMethodAction(call: call) else {
            result(FlutterMethodNotImplemented)
            return
        }
        
        let actionResult = methodAction.action(method: call)
        
        switch actionResult {
        case let successResult as SuccessResult:
            result(successResult.result)
        case let failureResult as FailureResult:
            result(FlutterError(code: failureResult.error, message: nil, details: nil))
        default:
            result(FlutterMethodNotImplemented)
        }
    }
}
